File: projects/api-models/skyportal_py_models/data_access_requests.py
"""Request and response models for SkyPortal data access requests."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

MAX_DATA_ACCESS_REQUESTS = 500


class DataAccessRequestPostBody(BaseModel):
    """Request body for asking an owner for data on an object."""

    model_config = ConfigDict(extra="forbid")

    objId: str = Field(description="ID of the object the data is attached to")
    photometry: list[PhotometryDataset] = Field(
        default_factory=list,
        description="Photometry datasets being asked for, as returned by the "
        "data availability endpoint.",
    )
    spectrumIDs: list[int] = Field(
        default_factory=list, description="IDs of the spectra being asked for"
    )
    message: str | None = Field(
        default=None, description="Note to the owner explaining the request"
    )


class DataAccessRequestPatchBody(BaseModel):
    """Request body for answering a request."""

    model_config = ConfigDict(extra="forbid")

    status: str = Field(description="Either 'accepted' or 'declined'")
    groupID: int | None = Field(
        default=None,
        description="Group to share the data into when accepting. Defaults to "
        "the requester's single user group.",
    )


class DataAccessRequestGetQuery(BaseModel):
    """Query parameters for listing data access requests."""

    model_config = ConfigDict(extra="forbid")

    objId: str | None = Field(default=None, description="Only requests on this object")
    status: str | None = Field(
        default=None, description="Only requests with this status"
    )
    direction: str | None = Field(
        default=None,
        description="'incoming' for requests to answer, 'outgoing' for requests "
        "made by the calling user. Both when omitted.",
    )
    pageNumber: int = Field(
        default=1,
        description="Page number for paginated query results. Defaults to 1.",
    )
    numPerPage: int = Field(
        default=25,
        description=(
            "Number of requests to return per paginated request. Defaults to "
            f"25. Max {MAX_DATA_ACCESS_REQUESTS}."
        ),
    )


MAX_DATA_ACCESS_REQUESTS = 500


class DataAccessRequestPostBody(BaseModel):
    """Request body for asking an owner for data on an object."""

    model_config = ConfigDict(extra="forbid")

    objId: str = Field(description="ID of the object the data is attached to")
    photometry: list[PhotometryDataset] = Field(
        default_factory=list,
        description="Photometry datasets being asked for, as returned by the "
        "data availability endpoint.",
    )
    spectrumIDs: list[int] = Field(
        default_factory=list, description="IDs of the spectra being asked for"
    )
    message: str | None = Field(
        default=None, description="Note to the owner explaining the request"
    )


class DataAccessRequestPatchBody(BaseModel):
    """Request body for answering a request."""

    model_config = ConfigDict(extra="forbid")

    status: str = Field(description="Either 'accepted' or 'declined'")
    groupID: int | None = Field(
        default=None,
        description="Group to share the data into when accepting. Defaults to "
        "the requester's single user group.",
    )


class DataAccessRequestGetQuery(BaseModel):
    """Query parameters for listing data access requests."""

    model_config = ConfigDict(extra="forbid")

    objId: str | None = Field(default=None, description="Only requests on this object")
    status: str | None = Field(
        default=None, description="Only requests with this status"
    )
    direction: str | None = Field(
        default=None,
        description="'incoming' for requests to answer, 'outgoing' for requests "
        "made by the calling user. Both when omitted.",
    )
    pageNumber: int = Field(
        default=1,
        description="Page number for paginated query results. Defaults to 1.",
    )
    numPerPage: int = Field(
        default=25,
        description=(
            "Number of requests to return per paginated request. Defaults to "
            f"25. Max {MAX_DATA_ACCESS_REQUESTS}."
        ),
    )


MAX_DATA_ACCESS_REQUESTS = 500


class DataAccessRequestGetQuery(BaseModel):
    """Query parameters for listing data access requests."""

    model_config = ConfigDict(extra="forbid")

    objId: str | None = Field(default=None, description="Only requests on this object")
    status: str | None = Field(
        default=None, description="Only requests with this status"
    )
    direction: str | None = Field(
        default=None,
        description="'incoming' for requests to answer, 'outgoing' for requests "
        "made by the calling user. Both when omitted.",
    )
    pageNumber: int = Field(
        default=1,
        description="Page number for paginated query results. Defaults to 1.",
    )
    numPerPage: int = Field(
        default=25,
        description=(
            "Number of requests to return per paginated request. Defaults to "
            f"25. Max {MAX_DATA_ACCESS_REQUESTS}."
        ),
    )


MAX_DATA_ACCESS_REQUESTS = 500


class PhotometryDataset(BaseModel):
    """One owner's photometry on an object in a single instrument/filter."""

    model_config = ConfigDict(extra="forbid")

    ownerID: int = Field(description="ID of the User who owns the photometry")
    instrumentID: int = Field(description="ID of the instrument it was taken with")
    filter: str = Field(description="Bandpass the photometry was taken in")


class DataAccessRequestPostBody(BaseModel):
    """Request body for asking an owner for data on an object."""

    model_config = ConfigDict(extra="forbid")

    objId: str = Field(description="ID of the object the data is attached to")
    photometry: list[PhotometryDataset] = Field(
        default_factory=list,
        description="Photometry datasets being asked for, as returned by the "
        "data availability endpoint.",
    )
    spectrumIDs: list[int] = Field(
        default_factory=list, description="IDs of the spectra being asked for"
    )
    message: str | None = Field(
        default=None, description="Note to the owner explaining the request"
    )


class DataAccessRequestPatchBody(BaseModel):
    """Request body for answering a request."""

    model_config = ConfigDict(extra="forbid")

    status: str = Field(description="Either 'accepted' or 'declined'")
    groupID: int | None = Field(
        default=None,
        description="Group to share the data into when accepting. Defaults to "
        "the requester's single user group.",
    )


class DataAccessRequestGetQuery(BaseModel):
    """Query parameters for listing data access requests."""

    model_config = ConfigDict(extra="forbid")

    objId: str | None = Field(default=None, description="Only requests on this object")
    status: str | None = Field(
        default=None, description="Only requests with this status"
    )
    direction: str | None = Field(
        default=None,
        description="'incoming' for requests to answer, 'outgoing' for requests "
        "made by the calling user. Both when omitted.",
    )
    pageNumber: int = Field(
        default=1,
        description="Page number for paginated query results. Defaults to 1.",
    )
    numPerPage: int = Field(
        default=25,
        description=(
            "Number of requests to return per paginated request. Defaults to "
            f"25. Max {MAX_DATA_ACCESS_REQUESTS}."
        ),
    )

File: static/js/ducks/dataAccessRequests.ts
import { skyportalApi } from "../api/skyportalApi";
import { invalidateOnMessage, findCachedQueryArg } from "../api/wsInvalidation";
import { dataAvailabilityTag } from "./dataAvailabilityTags";
import { sourceTag } from "./sourceTags";

export interface DataOwner {
  id: number;
  username: string;
  first_name: string | null;
  last_name: string | null;
}

export interface DataRequestStub {
  id: number;
  status: "pending" | "accepted" | "declined";
}

/** One owner's photometry on a source, in a single instrument and filter. */
export interface PhotometryAvailability {
  owner: DataOwner | null;
  instrument: { id: number; name: string } | null;
  filter: string;
  num_points: number;
  first_mjd: number | null;
  last_mjd: number | null;
  request: DataRequestStub | null;
}

export interface SpectrumAvailability {
  id: number;
  owner: DataOwner | null;
  instrument: { id: number; name: string } | null;
  observed_at: string | null;
  observed_at_mjd: number | null;
  type: string | null;
  label: string | null;
  origin: string | null;
  request: DataRequestStub | null;
}

export interface DataAvailability {
  photometry: PhotometryAvailability[];
  spectra: SpectrumAvailability[];
}

export interface DataAccessRequest {
  id: number;
  status: "pending" | "accepted" | "declined";
  message: string | null;
  obj_id: string;
  data_type: "photometry" | "spectrum";
  instrument_id: number | null;
  filter: string | null;
  spectrum_id: number | null;
  granted_group_id: number | null;
  created_at: string;
  requester: DataOwner;
  owner: DataOwner;
  /** Groups both the requester and the viewer belong to. */
  shareable_groups: { id: number; name: string }[];
}

export interface DataAccessRequestPage {
  requests: DataAccessRequest[];
  totalMatches: number;
  pageNumber: number;
  numPerPage: number;
}

export interface PhotometryDatasetRef {
  ownerID: number;
  instrumentID: number;
  filter: string;
}

export const dataAccessRequestsApi = skyportalApi.injectEndpoints({
  endpoints: (build) => ({
    getDataAvailability: build.query<DataAvailability, string>({
      query: (obj_id) => `api/sources/${obj_id}/data_availability`,
      providesTags: (_result, _error, obj_id) => dataAvailabilityTag(obj_id),
    }),
    getDataAccessRequests: build.query<
      DataAccessRequestPage,
      {
        direction?: "incoming" | "outgoing";
        status?: string;
        pageNumber?: number;
        numPerPage?: number;
      } | void
    >({
      query: (params) => ({
        url: "api/data_access_request",
        params: (params ?? {}) as Record<string, string>,
      }),
      providesTags: ["DataAccessRequest"],
    }),
    requestDataAccess: build.mutation<
      { ids: number[] },
      {
        objId: string;
        photometry?: PhotometryDatasetRef[];
        spectrumIDs?: number[];
        message?: string | null;
      }
    >({
      query: (body) => ({
        url: "api/data_access_request",
        method: "POST",
        body,
      }),
      invalidatesTags: (_result, _error, { objId }) => [
        ...dataAvailabilityTag(objId),
        "DataAccessRequest",
        ...sourceTag(objId),
      ],
    }),
    answerDataAccessRequest: build.mutation<
      unknown,
      { id: number; status: "accepted" | "declined"; groupID?: number | null }
    >({
      query: ({ id, ...body }) => ({
        url: `api/data_access_request/${id}`,
        method: "PATCH",
        body,
      }),
      invalidatesTags: ["DataAccessRequest", "DataAvailability"],
    }),
    withdrawDataAccessRequest: build.mutation<unknown, number>({
      query: (id) => ({
        url: `api/data_access_request/${id}`,
        method: "DELETE",
      }),
      invalidatesTags: ["DataAccessRequest", "DataAvailability"],
    }),
  }),
});

// The owner answering a request is a different session from the requester's, so
// the requester's list is refreshed by its own push.
invalidateOnMessage("skyportal/REFRESH_DATA_ACCESS_REQUESTS", () => [
  "DataAccessRequest",
]);

// A grant lands as new photometry/spectra on the source the requester is
// looking at. REFRESH_SOURCE is broadcast to every client for every source, and
// carries the source's internal_key, so translate that to the obj id and
// refresh only that source's availability. Without an obj id there is nothing
// to refresh: an unscoped tag here matches every open source page.
invalidateOnMessage("skyportal/REFRESH_SOURCE", (payload, getState) => {
  const objKey = payload?.obj_key as string | undefined;
  if (!objKey) {
    return null;
  }
  const objId = findCachedQueryArg(
    getState,
    "getSource",
    (data) => data?.internal_key === objKey,
  ) as string | null;
  return objId != null ? dataAvailabilityTag(objId) : null;
});

export const {
  useGetDataAvailabilityQuery,
  useGetDataAccessRequestsQuery,
  useRequestDataAccessMutation,
  useAnswerDataAccessRequestMutation,
  useWithdrawDataAccessRequestMutation,
} = dataAccessRequestsApi;
